webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,231 @@
+import { Button, InputNumber, Modal } from 'antd';
+import { Form, Input, Select } from 'antd';
+import { ModalProps } from 'antd/es/modal/Modal';
+import yaml from 'js-yaml';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import useStorage from 'hooks/useStorage';
+import { getResourcePools, getTaskTemplates } from 'services/api';
+import { NotebookConfig, RawJson, ResourcePool, ResourceType, Template } from 'types';
+import { launchNotebook, previewNotebook } from 'utils/task';
+
+import Link from './Link';
+import css from './NotebookModal.module.scss';
+import RadioGroup from './RadioGroup';
+import Spinner from './Spinner';
+
+const MonacoEditor = React.lazy(() => import('react-monaco-editor'));
+
+const { Option } = Select;
+const { Item } = Form;
+
+const STORAGE_PATH = 'notebook-launch';
+const STORAGE_KEY = 'notebook-config';
+
+interface Props extends ModalProps {
+  onLaunch?: () => void;
+  visible?: boolean;
+}
+
+const NotebookModal: React.FC<Props> = (
+  { visible = false, onLaunch, ...props }: Props,
+) => {
+  const storage = useStorage(STORAGE_PATH);
+  const [ showFullConfig, setShowFullConfig ] = useState(false);
+  const [ templates, setTemplates ] = useState<Template[]>([]);
+  const [ resourcePools, setResourcePools ] = useState<ResourcePool[]>([]);
+  const [ showResourceType, setShowResourceType ] = useState(true);
+  const [ resourceType, setResourceType ] = useState<ResourceType | undefined>
+  (storage.getWithDefault(STORAGE_KEY, { type: undefined }).type);
+  const [ form ] = Form.useForm();
+
+  const fetchTemplates = useCallback(async () => {
+    try {
+      setTemplates(await getTaskTemplates({}));
+    } catch {}
+  },[]);
+
+  useEffect(() => {
+    fetchTemplates();
+  }, [ fetchTemplates ]);
+
+  const fetchResourcePools = useCallback(async () => {
+    try {
+      setResourcePools( await getResourcePools({}));
+    } catch {}
+  }, []);
+
+  useEffect(() => {
+    fetchResourcePools();
+  }, [ fetchResourcePools ]);
+
+  const fetchConfig = useCallback(async () => {
+    try {
+      const values: NotebookConfig = form.getFieldsValue(true);
+      const config = await previewNotebook(
+        values.type === ResourceType.CPU ? 0 : values.slots,
+        values.template === 'default'? undefined : values.template,
+        values.name,
+        values.pool,
+      );
+      form.setFieldsValue({ config: yaml.dump(config) });
+    } catch {}
+  }, [ form ]);
+
+  useEffect(()=> {
+    if (showFullConfig){
+      fetchConfig();
+    }
+  }, [ showFullConfig, fetchConfig ]);
+
+  const storeConfig = useCallback((_, values: NotebookConfig) => {
+    delete values.name;
+    storage.set(STORAGE_KEY,values);
+  }, [ storage ]);
+
+  const handleSecondary = useCallback(() => {
+    if (showFullConfig) {
+      setShowFullConfig(show => !show);
+    } else {
+      form.validateFields().then(() => setShowFullConfig(show => !show)).catch();
+    }
+  },[ form, showFullConfig ]);
```

Comment:
**suggestion:** rewrite as an `async/await` form:
```
const handleSecondary = useCallback(async () => {
  if (showFullConfig) {
    setShowFullConfig(false);
  } else {
    try {
      await form.validateFields();
      setShowFullConfig(true);
    } catch (e) {}
  }
}, [ form, showFullConfig ]);
```
Also note that we don't need to check the previous state of the `showFullConfig` variable because we already checked if it was true or false.